docs(madr): logging standard (#4445)
kumahq/kuma: docs/madr/decisions/003-logging.md

## Decision Outcome

Chosen option: "Adopting consistent logging standard", because it's the only option.

### Positive Consequences

* Better UX for users.
* Better troubleshooting means both users and maintainers spend less time debugging issues.

### Negative Consequences

* More effort to try to follow the guidelines.

## Pros and Cons of the Options

### Adopting consistent logging standard

#### Convention

Ideally, all those points are linted.

1) Logs are lowercase. Can be uppercase when you start with proper noun.
_GOOD_
```go
log.Info("something has happened")
log.Info("Envoy is starting")
```
_BAD_
```go
log.Info("Something has happened")
```

2) Do not include `.` at the end of the sentence if this is the only sentence in the log
_GOOD_
```go
log.Info("something has happened")
log.Info("something has happened. Restarting the process")
```
_BAD_
```go
log.Info("something has happened.")
```

3) Either log or return an error. Don't do both, unless you want to log more details on DEBUG level.
_GOOD_
```go
if err != nil {
	return errors.Wrap(err, "more info")
}
...
if err != nil {
	log.Error(err, "more info")
	return nil
}
...
if err != nil {

> 	log.V(1).Info("much more details about the error")

Reviewer: Should the level be `DEBUG` as mentioned on line [56](https://github.com/kumahq/kuma/pull/4445/files#diff-03543e676073151a7aba162f34cbd764dc512a896e8f7a5bf2880bea0fe0bb85R56)?

Reviewer: Ok, I've just read [this](https://github.com/kumahq/kuma/pull/4445/files#diff-03543e676073151a7aba162f34cbd764dc512a896e8f7a5bf2880bea0fe0bb85R134) and it's so confusing. Why do we have `V(1).Info` which actually means `Debug`?

Author: Yes, `log.V(1).Info` means logging on `debug` level. `log.V(0).Info` means logging on `info` level. `V(0)` is explicit.

Reviewer: For me it's confusing to read something like

```
log...Info(...)
```

and find out that it actually means "debug". I think we either base our "levels" on numbers (0, 1...) or on words (info, debug ...)

Reviewer 2: Maybe we can wrap the `go-logr` interface and restrict it to `Info` == `V(0).Info`, `Debug` == `V(1).Info`.

As for conditionally doing `V(1).WithValues` maybe we can add a `WithDebugValues` which checks if `V(1)` before calling `WithValues`? It'd at least cover the common case where the value is already evaluated.

Author: Added clarification on how to use loggers on different levels.

For the wrapper. There is a slight chance that we will need something custom to implement this https://github.com/kumahq/kuma/issues/4446

I'm slightly anxious that the wrapper would affect the performance
https://github.com/go-logr/logr/blob/master/logr.go#L231

Author: @michaelbeaumont if you feel strongly about this logger wrapper, please file a separate issue since it's a bit out of scope of this PR (improvement of API, not the standard itself).

    return errors.Wrap(err, "more info")
}
```
_BAD_
```go
if err != nil {
    log.Error(err, "more info")
	return errors.Wrap(err, "more info")
}
```

4) Use constant message and structured logging

The first argument to the logger should be a constant message.

_GOOD_
```go
log.Info("dp connected", "name", dpName)
```
_BAD_
```go
log.Info(fmt.Sprintf("dp %s connected", dpName))
log.Info(fn(dpName)))
```

5) Carry the same context to multiple logs using WithValues
_GOOD_
```go
log := logger.WithValues("proxy", dpName, "mesh", mesh)
log.Info("connected")
log.Info("reconciled")
log.Info("disconnected")
```
_BAD_
```go
logger.Info("connected", proxy", dpName, "mesh", mesh)
logger.Info("reconciled", proxy", dpName, "mesh", mesh)
logger.Info("disconnected", proxy", dpName, "mesh", mesh)
```

6) Be mindful about logger name. Try to not repeat the information.
_GOOD_
```go
var log = core.Log.WithName("xds-server")
log.Info("starting")
```
_BAD_
```go
var log = core.Log.WithName("xds-server")
log.Info("starting XDS server")
```

7) Don't use ellipsis
_GOOD_
```go
log.Info("starting")
```
_BAD_
```go
log.Info("starting...")
```

> #### Levels

Reviewer: The full blown 7 levels (trace, debug, info, notice, warn, error, fatal) is usually an overkill but I'm a bit afraid this goes into the opposite direction. I agree with the use of `DEBUG`/`INFO` but I think we're missing `WARN`/`ERROR`.

From an operational perspective `WARN` (in my opinion) should be used for things that can **eventually cause problems but are not an immediate threat** e.g. a synchronisation mechanisms failed once or twice. It's not a problem now and it will probably fix itself (there might have been a short connectivity issue) but if this happens too many times you might end up in trouble. Logs with `WARN` should be reviewed at the end of the `on-call` shift, before handover.

`ERROR` should be used for issues that require immediate attention (or something close to immediate) e.g. the request to external service "score users trustworthiness" failed and we are exposed to fraud. This is usually coupled with metrics / alerts that page the engineer on-call. Should be reviewed as they happen / daily.

(disclaimer: not giving Kuma specific examples since I was struggling to find a good one, so I'm assuming there are examples like that)

---

With this MADR I'm thinking about the experience of a team running Kuma in production. I think we should add a point of being very explicit in the messages produced on the levels that are read by engineers outside of this team. We should not require people reading these logs to know any internal architecture / abbreviations etc.

(A bit off-topic but I think still useful to state) I think that `consul` does a good job of showing a "10000 foot view" of the [health of the cluster](https://learn.hashicorp.com/tutorials/consul/monitor-datacenter-health#autopilot). In their metrics it's called `autopilot` and it shows at a glance if the cluster is healthy or not.

Reviewer 3: Maybe worth adding that we want to follow the philosophy from this article

Author: > but I think we're missing WARN/ERROR

My bad. There is an error level in our logger. I took it for granted. However, we don't have WARN level.

> From an operational perspective WARN (in my opinion) should be used for things that can eventually cause problems but are not an immediate threat e.g. a synchronisation mechanisms failed once or twice. It's not a problem now and it will probably fix itself (there might have been a short connectivity issue) but if this happens too many times you might end up in trouble. Logs with WARN should be reviewed at the end of the on-call shift, before handover.

Sort of makes sense, but that sounds hard to implement. A couple of examples:
* XDS generation fails, should this be WARN or ERROR? It will be retried with watchdog, so should we keep a counter and after X times log error, not warn?
* Let say we receive a request with POST and postgres fails for a second. Should we do WARN or ERROR? A user most likely retries this. Does it require an immediate reaction?
* KDS resource exchange fails because of connectivity. Again, do we WARN and then after X time we log ERROR?

Is all of "after X time" hardcoded? Shouldn't this be a job for alerting system take into account thresholds for alerting of failed actions? Some users may want to know about it earlier, some later, because of different toleration for unavailability.

> I think it's following the idea behind: https://dave.cheney.net/2015/11/05/lets-talk-about-logging

Not exactly. This article takes the idea of dropping log levels to the extreme - you just need INFO and DEBUG.
I'd rather not link it.

> I think we should add a point of being very explicit in the messages produced on the levels that are read by engineers outside of this team. We should not require people reading these logs to know any internal architecture / abbreviations etc.

I tried to say that in the MADR. INFO level - need to be understandable without knowing implementation. DEBUG - requires impl knowledge.

Author: We discussed more offline with Krzysztof. The main problem was the lack of ERROR level which was my mistake in not describing it. I added it to the document.

The clear rules of ERROR vs INFO (=WARN in Kuma) are still fuzzy and IMHO hard to define as a set of rules.

#### Log or emit a metric

Metrics and logs are closely related, yet they have different characteristics. 
Logs give you the best precision. You know exactly what happened at specific time.
Logs usually give you more information than metrics. For example "User X accessed endpoint /xyz".
Metrics are more generalized so "Someone accessed endpoint /xyz".
In majority of cases, (structured) logging is superior, but it comes at cost of maintain the infrastructure to collect them.

For example, you may want to log every GET request to your server.
With systems that take a significant load, you may easily DDoS your logging system.
With metrics, you can increase counters and let the scraper fetch them every X second.

If you want to introduce a log that can be noisy, consider adding a generalized metric.
This way even if the user disables this log, they still have some data to troubleshoot. 

#### Global variable or pass the logger

In Kuma, we can either declare logger globally in the package
```go
var logger = core.Log.WithName("xds")
```
or pass it explicitly to the component
```go
NewComponent(core.Log.WithName("xds"))
```

> The library code should ALWAYS receive the logger. The user of the library should decide whether they want to log or not.

Reviewer: `library code` - what do you mean by that? is that packages in Kuma that could be reused somewhere else?

Author: Like I said in the MADR, I'd consider `pkg/core` and `pkg/utils` a library code. However, it is not explicit.
`pkg/core` is used heavily as library code in Kong Mesh which is built on top of Kuma.

We did the mistake of using `pkg` as the "default" directory for code instead of `internal` which seems to be a pattern in more modern Go projects. This way `internal` is internal code and `pkg` is an explicit library code.

# Logging standard

* Status: accepted

## Context and Problem Statement

Seamless troubleshooting is a feature. We should keep more attention to the logging in Kuma.

## Considered Options

* Adopting consistent logging standard.

## Decision Outcome

Chosen option: "Adopting consistent logging standard", because it's the only option.

### Positive Consequences

* Better UX for users.

> * Better troubleshooting means both users and maintainers spent less time debugging an issue.

Reviewer 3: ```suggestion
* Better troubleshooting means both users and maintainers spend less time debugging issues.
```

### Negative Consequences

* More effort to try to follow the guidelines.

## Pros and Cons of the Options

### Adopting consistent logging standard

#### Convention

> Ideally, all those points are linted.

Reviewer 3: Seems like it's technically feasible: https://golangci-lint.run/contributing/new-linters/

Author: https://github.com/kumahq/kuma/issues/4513 created issue

1) Logs are lowercase. Can be uppercase when you start with proper noun.
_GOOD_
```go
log.Info("something has happened")
log.Info("Envoy is starting")
```
_BAD_
```go
log.Info("Something has happened")
```

2) Do not include `.` at the end of the sentence if this is the only sentence in the log
_GOOD_
```go
log.Info("something has happened")
log.Info("something has happened. Restarting the process")
```
_BAD_
```go
log.Info("something has happened.")
```

3) Either log or return an error. Don't do both, unless you want to log more details on DEBUG level.
_GOOD_
```go
if err != nil {
	return errors.Wrap(err, "more info")
}
...
if err != nil {
	log.Error(err, "more info")
	return nil
}
...
if err != nil {
	log.V(1).Info("much more details about the error")
    return errors.Wrap(err, "more info")
}
```
_BAD_
```go
if err != nil {
    log.Error(err, "more info")
	return errors.Wrap(err, "more info")
}
```

> 4) Use structured logging instead of `fmt.Sprintf`

Reviewer 3: Maybe this should be even saying: "the first arg should always be a constant string". String concat, function calls etc should be used

_GOOD_
```go
log.Info("dp connected", "name", dpName)
```
_BAD_
```go
log.Info(fmt.Sprintf("dp %s connected", dpName))
```

5) Carry the same context to multiple logs using WithValues
_GOOD_
```go
log := logger.WithValues("proxy", dpName, "mesh", mesh)
log.Info("connected")
log.Info("reconciled")
log.Info("disconnected")
```
_BAD_
```go
logger.Info("connected", proxy", dpName, "mesh", mesh)
logger.Info("reconciled", proxy", dpName, "mesh", mesh)
logger.Info("disconnected", proxy", dpName, "mesh", mesh)
```

6) Be mindful about logger name. Try to not repeat the information.
_GOOD_
```go
var log = core.Log.WithName("xds-server")
log.Info("starting")
```
_BAD_
```go
var log = core.Log.WithName("xds-server")
log.Info("starting XDS server")
```

7) Don't use ellipsis
_GOOD_
```go
log.Info("starting")
```
_BAD_
```go
log.Info("starting...")
```

> 8) Log before executing the action

Reviewer 2: Is there guidance about what "an action" is? Or is it just a "you know it when you see it" thing.

> Logging before executing the significant action is important because if there is an error, it's easy to see the initial cause.

Reviewer 2: ```suggestion
Logging before executing the significant action is important because we can see how long something takes.
```

May or may not have been obvious from the discussion but we should be clear the error cause isn't a reason to log before/after. The cause should come from the error itself, which we should in general always wrap to add context, independent of  how we log.
Wrapping errors consistently will make it more often unnecessary to even look through the logs.
